src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
+
+> **Edition Differences**: The `?` repetition operator did not exist before the
+> 2018 edition. Prior to the 2018 Edition, `?` was an allowed
+> separator token, rather than a repetition operator.
+
+## Scoping, Exporting, and Importing
+
+For historical reasons, the scoping of macros by example does not work entirely like
+items. Macros have two forms of scope: textual scope, and path-based scope.
+Textual scope is based on the order that things appear in source files, or even
+across multiple files, and is the default scoping. It's explained further below.
+Path-based scope works exactly the same way that item scoping does. The scoping,
+exporting, and importing of macros is controlled largely by attributes.
+
+When a macro is invoked by an unqualified identifier (not part of a multi-part
+path), it's first looked up in textual scoping. If this does not yield any
+results, then it is looked up in path-based scoping. If the macro's name is
+qualified with a path, then it is only looked up in path-based scoping.
+
+```ignore
+use lazy_static::lazy_static; // Path-based import.
+
+macro_rules! lazy_static { // Textual definition.
+    (lazy) => {};
+}
+
+lazy_static!{lazy} // Textual lookup finds our macro first.
+self::lazy_static!{} // Path-based lookup ignores our macro, finds imported one.
+```
+
+### Textual Scope
+
+Textual scope is based largely on the order that things appear in source files.
+When `macro_rules!` is used to define a macro, the macro enters scope after the
+definition (note that it can still be used recursively, since names are looked
+up from the invocation site), up until its module is closed. This can enter
+child modules and even span across multiple files:
+
+```ignore
+//// src/lib.rs
+mod has_macro {
+    // m!{} // Error: m is not in scope.
+
+    macro_rules! m {
+        () => {};
+    }
+    m!{} // OK: appears after declaration of m.
+
+    mod uses_macro;
+}
+
+// m!{} // Error: m is not in scope.
+
+//// src/has_macro/uses_macro.rs
+
+m!{} // OK: appears after delcaration of m in src/lib.rs
+```
+
+It is not an error to define a macro multiple times; the most recent declaration
+will shadow the previous one unless it has gone out of scope.
+
+```
+macro_rules! m {
+    (1) => {};
+}
+
+m!(1);
+
+mod inner {
+    m!(1);
+
+    macro_rules! m {
+        (2) => {};
+    }
+    // m!(1); // Error: no rule matches '1'
+    m!(2);
+
+    macro_rules! m {
+        (3) => {};
+    }
+    m!(3);
+}
+
+m!(1);
+```
+
+The `#[macro_use]` attribute has two purposes. First, it can be used to make a
+module's macro scope not end when the module is closed, by applying it to a
+module:
+
+```
+#[macro_use]
+mod inner {
+    macro_rules! m {
+        () => {};
+    }
+}
+
+m!();
+```
+
+Second, it can be used to import macros from another crate, by attaching it to
+an  `extern crate` declaration. Macros imported this way are imported textually
+only. Optionally, a list of macros to import can be specified; this is not
```

Comment:
`#[macro_use]` on `extern crate` doesn't work textually (e.g. you can use the macro textually before the `extern crate` item).
It adds the macro into prelude instead.
(Also the `extern crate` item must be in the crate root, otherwise `#[macro_use]` is an error.)

The only "textual" aspect here is that in
```rust
#[macro_use]
extern crate a;
#[macro_use]
extern crate b;
```
prelude entries introduced by `b` will override entries previously introduced by `a`.

This is quite unfortunate, but understandable from backward-compatibility point of view - before Rust 1.29 if you had and ambiguity between macros provided by `a` and `b` there was no way to explicitly disambiguate, so we couldn't report ambiguity errors in this situation.